Update existing notification pattern with new styling

Caleb-Ellis opened the pull request:
## Done

- Updated existing notification pattern to use new styling

Fixes #3822 

## QA

- Open [demo](https://vanilla-framework-3844.demos.haus/docs/examples/patterns/notifications/caution)
- Check that it matches the new [design](https://app.zeplin.io/project/5fac06865cfff51271b72cc6/screen/60353ac1f130133007d40217)

### Check if PR is ready for release

If this PR contains Vanilla SCSS code changes, it should contain the following changes to make sure it's ready for the release:

- [x] PR should have one of the following labels to automatically categorise it in release notes:
  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
- [x] Vanilla version in `package.json` should be updated relative to the [most recent release](https://github.com/canonical-web-and-design/vanilla-framework/releases/latest), following semver convention:
  - if CSS class names are not changed it can be bugfix relesase (x.x.**X**)
  - if CSS class names are changed/added/removed it should be minor version (x.**X**.0)
  - see the [wiki for more details](https://github.com/canonical-web-and-design/vanilla-framework/wiki/Release-process#pre-release-tasks)
- [x] Any changes to component class names (new patterns, variants, removed or added features) should be listed on the [component status page](https://github.com/canonical-web-and-design/vanilla-framework/blob/master/templates/docs/component-status.md).
- [x] Documentation side navigation should be updated with the relevant labels.


## Screenshots

![2021-07-05_12-06](https://user-images.githubusercontent.com/25733845/124408392-7ea60680-dd89-11eb-9aa0-54ccc03ed7a7.png)


## Questions
- Is this pattern meant for a minor or major release?
- Should we update the documentation immediately or when the pattern is finished?
- What should the default style be that would show if someone updated Vanilla without changing the markup at all? Should it be full-width?
- Is the notification icon always linked to the highlight bar, or should they be able to be changed independently?
- Is the default notification supposed to use the warning icon?
- Is the "date" section of the new notification only allowed to be dates? Or should it be able to be any supplementary text (in other words, can it potentially be really long)?
- The design shows that the bottom-right can contain action "links", action "buttons", a contextual menu or a "view more notifications" text. Are they mutually-exclusive? If not, how should they display when there are multiple? Also, what are the circumstances in which you'd use an action "link" over an action "button"?
- Are the stacked notifications maxed out at three with the extras hidden?
- Can a stacked notification also have the overlapping or full-width styles?
- When expanding the stacked notifications, do they just all show one on top of the other? Is there animation?
- Do we want to show each stacked notification with the real highlight colour? Or can we "fake" it to just show the same as the top notification if we're assuming stacked notifications will only ever be grouped by type.

webteam-app commented:
Demo starting at https://vanilla-framework-3844.demos.haus

bartaz commented:
> * Is this pattern meant for a minor or major release?

I guess you mean patch (_._.X) or  minor (_.X._)? We generally tend to avoid breaking changes as much as possible and only release major (X._._) rarely.

In the case of this PR it would be patch release. We are not planning to change anything about "API" which in Vanilla case is CSS class names. As long as we don't introduce new CSS class names or deprecate old ones we release patch versions.

Later on, when you start working on new updated markup for notifications, this will be a minor release (as it will introduce new CSS class names quite).

> * Should we update the documentation immediately or when the pattern is finished?

We try to make our PRs ready for release, so any necessary documentation changes would be included.
In case of this PR I don't think any would be necessary, as it's meant to be only style changes for existing component.


> * What should the default style be that would show if someone updated Vanilla without changing the markup at all? Should it be full-width?
> * Is the notification icon always linked to the highlight bar, or should they be able to be changed independently?
> * Is the default notification supposed to use the warning icon?
> * Is the "date" section of the new notification only allowed to be dates? Or should it be able to be any supplementary text (in other words, can it potentially be really long)?
> * The design shows that the bottom-right can contain action "links", action "buttons", a contextual menu or a "view more notifications" text. Are they mutually-exclusive? If not, how should they display when there are multiple? Also, what are the circumstances in which you'd use an action "link" over an action "button"?
> * Are the stacked notifications maxed out at three with the extras hidden?
> * Can a stacked notification also have the overlapping or full-width styles?
> * When expanding the stacked notifications, do they just all show one on top of the other? Is there animation?
> * Do we want to show each stacked notification with the real highlight colour? Or can we "fake" it to just show the same as the top notification if we're assuming stacked notifications will only ever be grouped by type.

wgx commented:
>Is the "date" section of the new notification only allowed to be dates? Or should it be able to be any supplementary text (in other words, can it potentially be really long)?

It's designed to show recency (and therefore allow the user to imply urgency. The intention is it would only be a date. 

>The design shows that the bottom-right can contain action "links", action "buttons", a contextual menu or a "view more notifications" text. Are they mutually-exclusive? If not, how should they display when there are multiple? Also, what are the circumstances in which you'd use an action "link" over an action "button"?

This has changed since the wireframes, and I'm not sure. In the wires, all notificaitons had the option of the context menu because it's how you set a notification to `snooze` or `mark as read` - @lyubomir-popov 

>When expanding the stacked notifications, do they just all show one on top of the other? Is there animation?

Visual concern, but maybe we do show a little micro-animation. 

>Do we want to show each stacked notification with the real highlight colour? Or can we "fake" it to just show the same as the top notification if we're assuming stacked notifications will only ever be grouped by type.

I imagined that they'd be grouped by color/severity, but if we decide not to - we could show a 'grey' color bar when the contents of the stack is mixed. Or.. the most severe colour in the top of the stack?

lyubomir-popov commented:
@Caleb-Ellis 

line 2 needs to change from:
`$response-top-padding: $spv-nudge + $spv-inner--small;`
to:
$response-top-padding: $spv-inner--small;

Caleb-Ellis commented:
@bartaz @lyubomir-popov I've updated it so the spacing is tightened, also I've changed the default `.p-notification` class to give it the info styling

bartaz commented:
@Caleb-Ellis Great, thanks for updates. It seems that `p-icon--information` doesn't have icon or highlight now:

<img width="853" alt="Screenshot 2021-07-05 at 15 54 36" src="https://user-images.githubusercontent.com/83575/124482329-7231a600-dda9-11eb-99bc-0c691ce13cff.png">

For consistency we want to keep it as having the blue highlight and information icon. So `p-notification` and `p-notification--information` would be the same.

bartaz commented on scss/_patterns_notifications.scss:
```diff
@@ -110,6 +116,5 @@ $response-icon-space: 2 * $sph-inner + map-get($icon-sizes, default);
 @mixin vf-notifications-information {
   .p-notification--information {
     @extend %vf-notification;
-    @include vf-highlight-bar($color-information);
```
We should keep the blue highlight on the left here: 

```scss
@include vf-highlight-bar($color-information);
```

Also, information notification didn't have any icon previously, but for consistency it should also have information icon (same as default notification):

```scss
.p-notification__response {
    @include vf-icon-information;
}
```

bartaz reviewed (approved):
@Caleb-Ellis @huwshimi 

This looks good, thanks.
Just one last thing would be to make sure `p-notification--information` is styled properly (with blue highlight and info icon) as well. You can merge this one when ready.

I figured that one thing that may come in way is Percy. Once you push anything you will need to wait for it to run (it takes a bit, we have a lot of screenshots to process) and possibly approve whatever it notices. I think you both have access to Percy, so feel free to approve for yourself.
In Percy you may notice some weird unrelated screenshots with differences. It's possible, we have a couple that show up once in a while. Most notably: Firefox table spacing bug page, sometimes embedded media shows up with error (when youtube quota is exceeded by percy), and sometimes a button example shows up as a new snapshot (for unknown reason). Feel free to safely ignore any of these.
If you notice some other screenshot that you don't know why is flagged by percy, just throw us a link in mattermost of something, we can figure out later. Don't let it block you.

lyubomir-popov commented:
You can run locally to save time - also limit to only a subset of the images, iirc - details [here ](https://docs.percy.io/docs/local-development)
